core/src/main/java/org/geysermc/geyser/translator/protocol/java/JavaUpdateRecipesTranslator.java

```diff
@@ -214,23 +229,27 @@ public void translate(GeyserSession session, ClientboundUpdateRecipesPacket pack
             }
         }
 
-        // FIXME: if the server/viaversion doesn't send trim recipes then we shouldn't either.
-
-        // BDS sends armor trim templates and materials before the CraftingDataPacket
-        TrimDataPacket trimDataPacket = new TrimDataPacket();
-        trimDataPacket.getPatterns().addAll(TrimRecipe.PATTERNS);
-        trimDataPacket.getMaterials().addAll(TrimRecipe.MATERIALS);
-        session.sendUpstreamPacket(trimDataPacket);
+        session.getLastRecipeNetId().set(netId);
 
-        // Identical smithing_trim recipe sent by BDS that uses tag-descriptors, as the client seems to ignore the
-        // approach of using many default-descriptors (which we do for smithing_transform)
-        craftingDataPacket.getCraftingData().add(SmithingTrimRecipeData.of(TrimRecipe.ID,
-                TrimRecipe.BASE, TrimRecipe.ADDITION, TrimRecipe.TEMPLATE, "smithing_table", netId++));
+        // Only send smithing trim recipes if Java/ViaVersion sends them.
+        if (sendTrimRecipes) {
+            // BDS sends armor trim templates and materials before the CraftingDataPacket
+            TrimDataPacket trimDataPacket = new TrimDataPacket();
+            trimDataPacket.getPatterns().addAll(TrimRecipe.PATTERNS);
+            trimDataPacket.getMaterials().addAll(TrimRecipe.MATERIALS);
+            session.sendUpstreamPacket(trimDataPacket);
 
+            // Identical smithing_trim recipe sent by BDS that uses tag-descriptors, as the client seems to ignore the
+            // approach of using many default-descriptors (which we do for smithing_transform)
+            craftingDataPacket.getCraftingData().add(SmithingTrimRecipeData.of(TrimRecipe.ID,
+                    TrimRecipe.BASE, TrimRecipe.ADDITION, TrimRecipe.TEMPLATE, "smithing_table", session.getLastRecipeNetId().getAndIncrement()));
+        } else {
+            session.setOldSmithingTable(true);
+            craftingDataPacket.getCraftingData().addAll(getUpgradeRecipes(session));
+        }
         session.sendUpstreamPacket(craftingDataPacket);
         session.setCraftingRecipes(recipeMap);
         session.setStonecutterRecipes(stonecutterRecipeMap);
-        session.getLastRecipeNetId().set(netId);
```

Comment:
?

Reply:
https://github.com/GeyserMC/Geyser/pull/4056/files#diff-63970a81cc5c7b300e04453fd703f6292c88875c69b956dc48ef5a7793f8bea9R232 Had to move it up; otherwise, getUpgradeRecipes can't get the latest value of netId to get and increment from